scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -106,198 +113,198 @@ public SortedBucketSource(
       Class<FinalKeyT> finalKeyClass,
       List<BucketedInput<?, ?>> sources,
       TargetParallelism targetParallelism) {
+    // Initialize with absolute minimal parallelism and allow split() to create parallelism
+    this(finalKeyClass, sources, targetParallelism, 0, 1);
+  }
+
+  private SortedBucketSource(
+      Class<FinalKeyT> finalKeyClass,
+      List<BucketedInput<?, ?>> sources,
+      TargetParallelism targetParallelism,
+      int bucketOffsetId,
+      int effectiveParallelism) {
     this.finalKeyClass = finalKeyClass;
     this.sources = sources;
     this.targetParallelism = targetParallelism;
+    this.bucketOffsetId = bucketOffsetId;
+    this.effectiveParallelism = effectiveParallelism;
+    this.keyGroupSize =
+        Metrics.distribution(SortedBucketSource.class, "SortedBucketSource-KeyGroupSize");
```

Comment:
Unlike in the previous iteration where `SortedBucketSource` is a PTransform, we don't have access to a unique name for this Read so if there are multiple SortedBucketSource reads in a single pipeline the counters would overlap... any ideas on a way to disambiguate them?

Reply:
I added an optional metricsKey argument that will default to `SortedBucketSource-{$atomicInteger}` if not supplied. If `atomicInteger == 0` it will just be `SortedBucketSource`... so for people who only have a single SMB read in their pipeline, the metrics should look exactly the same.